watchtower/wtdb/tower_db_test.go

```diff
@@ -0,0 +1,712 @@
+	setAndCheck := func(i int) {
+		expEpoch := epochFromInt(1)
+		err = h.db.SetLookoutTip(expEpoch)
+		if err != nil {
+			h.t.Fatalf("unable to set lookout tip: %v", err)
+		}
+
+		epoch, err = h.db.GetLookoutTip()
+		if err != nil {
+			h.t.Fatalf("unable to fetch lookout tip: %v", err)
+		}
+
+		if !reflect.DeepEqual(epoch, expEpoch) {
+			h.t.Fatalf("lookout tip mismatch, want: %v, got: %v",
+				expEpoch, epoch)
+		}
+	}
+
+	// Set and assert the lookout tip.
+	for i := 0; i < 5; i++ {
+		setAndCheck(i)
+	}
+}
+
+// testDeleteSession asserts the behavior of a tower database when deleting
+// session data. The test asserts that the only proper the target session is
+// remmoved, and that only updates for a particular session are pruned.
+func testDeleteSession(h *towerDBHarness) {
+	// First, create a session so that the database is not empty.
+	id0 := id(0)
+	session0 := &wtdb.SessionInfo{
+		ID: *id0,
+		Policy: wtpolicy.Policy{
+			MaxUpdates: 3,
+		},
+		RewardAddress: []byte{},
+	}
+	h.insertSession(session0, nil)
+
+	// Now, attempt to delete a session which does not exist, that is also
+	// different from the first one created.
+	id1 := id(1)
+	h.deleteSession(*id1, wtdb.ErrSessionNotFound)
+
+	// The first session should still be present.
+	h.getSession(id0, nil)
+
+	// Now insert a second session under a different id.
+	session1 := &wtdb.SessionInfo{
+		ID: *id1,
+		Policy: wtpolicy.Policy{
+			MaxUpdates: 3,
+		},
+		RewardAddress: []byte{},
+	}
+	h.insertSession(session1, nil)
+
+	// Create and insert updates for both sessions that have the same hint.
+	var hint wtdb.BreachHint
+	update0 := &wtdb.SessionStateUpdate{
+		ID:            *id0,
+		Hint:          hint,
+		SeqNum:        1,
+		EncryptedBlob: []byte{},
+	}
+	update1 := &wtdb.SessionStateUpdate{
+		ID:            *id1,
+		Hint:          hint,
+		SeqNum:        1,
+		EncryptedBlob: []byte{},
+	}
+
+	// Insert both updates should succeed.
+	h.insertUpdate(update0, nil)
+	h.insertUpdate(update1, nil)
+
+	// Remove the new session, which should succeed.
+	h.deleteSession(*id1, nil)
+
+	// The first session should still be present.
+	h.getSession(id0, nil)
+
+	// The second session should be removed.
+	h.getSession(id1, wtdb.ErrSessionNotFound)
+
+	// Assert that only one update is still present.
+	matches := h.queryMatches(hint)
+	if len(matches) != 1 {
+		h.t.Fatalf("expected one update, found: %d", len(matches))
+	}
+
+	// Assert that the update belongs to the first session.
+	if matches[0].ID != *id0 {
+		h.t.Fatalf("expected match for %v, instead is for: %v",
+			*id0, matches[0].ID)
+	}
+
+	// Finally, remove the first session added.
+	h.deleteSession(*id0, nil)
+
+	// The session should no longer be present.
+	h.getSession(id0, wtdb.ErrSessionNotFound)
+
+	// No matches should exist for this hint.
+	matches = h.queryMatches(hint)
+	if len(matches) != 0 {
+		h.t.Fatalf("expected zero updates, found: %d", len(matches))
+	}
+}
+
+type stateUpdateTest = struct {
```

Comment:
Extraneous `=` here? Didn't even know something like this would compile lol

Reply:
heh that's strange. i suppose it's just an alias for an unnamed struct, which seems valid. can't think off the top of my head how that'd be useful on its own tho

fixed e098bd44ef8d66aaa0a94ee72d1fa710e88727b6